### ISSUE TITLE
Channel entities produce message entities of inconsistent types

### ISSUE THREAD

**Issue by kane-t** [RogueException/Discord.Net#890]
It's impossible to reliably obtain a `SocketMessage` from a `SocketChannel`.

Suppose I have a text channel (as `ISocketMessageChannel`), a message ID, and a `DiscordSocketClient`, and I want to retrieve the text message as a `SocketMessage`.  If I call `ISocketMessageChannel.GetMessageAsync()`, the result is an `IMessage`, not a `SocketMessage`, and looking at the source suggests that the actual concrete type of the object will be `SocketMessage` if the message was cached, otherwise it'll be `RestMessage`.  There's no way I can see to convert a `RestMessage` to a `SocketMessage`.

Since I can't reliably get `SocketMessages` from `SocketChannels` in all cases, there's little point to me ever using `SocketMessage` instead of `IMessage`.  It'll just cause problems when, for example, deserialising data from a previous session.  But that means that the `Author` property will always produce `IUser`s, which won't always be `SocketUser`s, so I'll have to be constantly calling `GetUser` on guilds to get user objects, instead of getting them from the message.  This is incredibly convoluted.

`ISocketMessageChannel` should have a full set of `GetMessageAsync` and `GetMessagesAsync` methods that return `SocketMessage`s.  The `IMessage`-returning functions from the IMessageChannel base class can just defer to these, since `SocketMessage` implements `IMessage`.  More generally, Socket entities should always produce Socket entities.  If I retrieve a user from a `SocketChannel`, it should be a `SocketUser`; if I retrieve a channel from a `SocketGuild`, it should be a `SocketChannel`; if I retrieve a message from a `SocketChannel`, it should be a `SocketMessage`.

**Comment by foxbot** [comment]
This would appear to be because you do not have the message cache enabled -- so the GetMessage(s)Async method will _always_ need to ping Discord to get a message. You can enable the message cache by setting `MessageCacheSize` in `DiscordSocketConfig`.

To force all objects returned from these calls to be a type of Socket entity, you can specify `CacheMode.CacheOnly` as a parameter to `GetMessage(s)Async`.

If you always need a `SocketUser` object, your best bet would be to do pattern matching on the returned messages:
```cs
SocketUser user = message is SocketMessage s ? s.Author : guild.GetUser(message.Author.Id);
```

**Comment by kane-t** [comment]
The message cache is enabled, but if the message isn't already cached when I call `GetMessageAsync`, it'll be retrieved and returned as a `RestMessage`, as visible here:
https://github.com/RogueException/Discord.Net/blob/1ce1c019b399b83fb7ed946ec7908d5d37b567fb/src/Discord.Net.Rest/Entities/Channels/ChannelHelper.cs#L97

If I set `CacheOnly`, won't that just return `null` if the message wasn't already cached?  It looks to me like `CacheOnly` means "only return a result if it's already in the cache."

This *may* be unintentional behaviour.  There's a `SocketChannelHelper`, but all `SocketChannel` entites call down to `ChannelHelper`, which always returns `RestMessages`.

**Comment by AntiTcb** [comment]
>More generally, Socket entities should always produce Socket entities.

Socket entities are largely cacheable because they are only able to be created when the gateway sends information about them. They cannot be retrieved and created on demand. As such, for consumer's sake, the socket entities will also encapsulate REST requests, otherwise, there'd be a lot of missing functionality such as not being able to send messages via `IMessageChannel` objects, basically anything involving invites, etc, without needing to also explicitly have a `DiscordRestClient` that you work with in addition to the `DiscordSocketClient`.

**Comment by kane-t** [comment]
Hmm, I see.  The WebSocket gateway doesn't provide a way to retrieve specific messages by ID?

Still, is there some specific reason that a `SocketMessage` can't just be created from a `RestMessage` after it's retrieved through the REST API?  All of the other entities inside a `SocketMessage` are ones that *do* seem to all be available and cached--users, roles, and channels.  The only difference in the class appears to be the `CreatedAt` field in SocketMessage, but that's just calculated from the snowflake, which is the same regardless of whether the message comes from the REST or WebSocket APIs.

You'd have to fill out the fields by retrieving the socket entities from the socket client, but I don't see an obvious problem with that.  And only you can do that, because all of the entities' constructors are private.

**Comment by Joe4evr** [comment]
`it's by design that we don't wait for messages to be received on the gateway before returning them as it could lead to deadlocks`

**Comment by foxbot** [comment]
Caching messages sent over the REST API would kind of create a race condition - we're arbitrarily inserting a message into the state, when that message could be in the process of being updated over the socket.

Realistically if you're pulling these messages from `SocketGuild.GetMessageAsync`, you already have them in the context of a guild - so just take the properties you want from them, and use your socket-attached entities for anything else.

We categorize entities by their source of origin: messages received over the WebSocket are SocketXXX entities, while messages received over the REST API are RestXXX entities. It wouldn't make sense for us to pull a detached message from the REST API and call it a Socket entity.

**Comment by foxbot** [comment]
Regarding your earlier question: no, we are very limited in what we can request over Discord's WebSocket.

They send us a very limited version of the state when we open a connection -- guilds, channels, and online members. The only additional data we can ever request over the socket is member data for members that are offline (`DiscordSocketConfig.DownloadAllUsers`) - and even this is discouraged, as its a lot of unnecessary bandwidth for Discord to deal with.

As far as messages are concerned, we only receive a message when it is sent or updated -- and in the case of updated messages, we don't receive a historical copy of it, only the most recent copy.

**Comment by kane-t** [comment]
Ah, I see.  So, what I'm getting is that I shouldn't be using `SocketMessage` for when I'm expecting to persist the data across disconnects and reconnects, because while users, roles, channels, and such, are all fully-populated by the gateway, messages only come over the WebSocket when they're newly-created or updated.  I should switch my code over to using `RestMessage` for cases where, for example, I need to serialise the message and then deserialise it to use later.

(For the record, the case in question is when the bot I'm writing shuts down while it has queued commands to execute.  In 0.9, I was just writing out the IDs of the command messages to a file when disconnecting, and then when I reconnected I read the message IDs back in, pulled the messages from the API, and stuck them back in the queue of commands waiting to be handled.  When I switched to 1.0, I changed that queue from a `Queue<Message>` to a `Queue<SocketMessage>`, which clearly won't work.)

Thanks!  You might also consider putting a note about that in the documentation--the docs as written seemed to imply to me that the REST version of the library was just there for compatibility, and the WebSocket client was the preferred choice.  It didn't occur to me that it'd be missing any functionality from the REST client.
